### ISSUE TITLE
Bandwidth Improvements

### ISSUE THREAD

**Issue by williambanfield** [tendermint/tendermint#9575]
This issue tracks the work associated with improving the bandwidth consumed by a Tendermint node running on a production network

As envisioned, this project can be broken down into 3 phases.

### 1. Data gathering

Before suggesting or implementing any improvements within Tendermint, it's critically important that we understand specifically which components of Tendermint are consuming a lot of bandwidth.

There is a set of prometheus metrics within Tendermint at our disposal to determine the bandwidth consumption:

```prometheus
p2p_peer_send_bytes_total
p2p_peer_receive_bytes_total
```
Each of these metrics is labeled with the ID of the corresponding peer and the ID of the Tendermint channel. Each channel ID maps to a different subcomponent of Tendermint and can therefore be used to determine which protocols or parts of the process are consuming large amounts of bandwidth. 

While useful, this data may be somewhat incomplete for our purposes. The protocols of Tendermint can send different message _types_ on each channel, each of which serve different purposes.  narrow down which specific action in the protocols is consuming the most bandwidth, we should further instrument the prometheus metrics with the message type. This was done in an alternate version of Tendermint, v0.36.x, and should be ported across the `main`.

- [x] reimplement #7155

Consider creating a new metric instead that just contains the message / channel type instead of adding an additional label to the existing metric. The existing metric already has a very high cardinality.

Once this metric change is re-implemented, we'll need to gather data on a real network with the metric in place. Since the primary network reporting an issue is osmosis, this network should be targeted first. We should contact a active validator and request that they run with the updated metric for a period of a few hours and provide us with a dump of the prometheus data collected in that time. Since the network operators report incredibly high bandwidth and tendermint's protocols are consistent over time, an observation period of a few hours should be enough to capture a representative sample. 

### 2. Code analysis and bandwidth improvement suggestions

Once phase 1. is complete, it will be clear which message types are consistently consuming the most bandwidth. We should use this information to inspect the code for functionality that contributor to the consumption. 

The Tendermint protocols are both 'chatty' and reasonably complex, so there may be no obvious ways to reduce bandwidth. If there are any clearly unnecessary sections that consume high bandwidth, they should be considered for update immediately. However, it's likely that no quick fixes are possible without either breaking the protocol or requiring a more complex, and therefore riskier fix. In these cases, the sections that consume large amounts of bandwidth should still be cataloged.

Information on which message types consume high amounts of bandwidth, which corresponding pieces of code are responsible for sending those messages, and possible updates should all be outlined in a brief report that can be used both for phase 3 of this project as well as by future projects that attempt to change Tendermint's bandwidth utilization.

- [x] #9576

### 3. Implementation of suggestions proposed in phase 2

Once the data collection and inspection is complete, we can actually begin making changes to Tendermint. Any possible quick fixes - obvious and small scale changes - can likely be performed right away without much discussion. Larger changes to protocols or Tendermint internals should be accompanied by an ADR describing how the intended change will reduce bandwidth and why it is safe. Any changes should aim to be non-breaking.

**Comment by sergio-mena** [comment]
Thanks @williambanfield, my comments:
* on 1. Even if (as the text says) we need to extend our metrics (actually, backport the metric extensions IIUC), and then re-run some kind of realistic setup, would it make sense to (as a preliminary step) take a closer look at the data we already gathered during the QA process?
* on 3. I think it would be good to implement the "obvious changes", if any, as soon as possible, even before or during the report write-up in 2.

**Comment by williambanfield** [comment]
1. Sure thing. I took a quick glance and a lot of it is on the consensus channels which each send a receive multiple kinds of messages so it's hard to tell what exactly is consuming the bandwidth.
3. Yes, I think that seems straightforward enough and makes sense.

**Comment by williambanfield** [comment]
Below is key that indicates which Tendermint channel each hex value in the following graphs corresponds to:
```
0x0:  PexChannel
0x20: StateChannel
0x21: DataChannel
0x22: VoteChannel
0x23: VoteSetBitsChannel
0x30: MempoolChannel
0x40: BlocksyncChannel
```

Peer Send Bytes:
![image](https://user-images.githubusercontent.com/4561443/197577775-ed0f97a6-354f-4971-acdb-0fa9837dc6cd.png)

Peer Receive Bytes:
![image](https://user-images.githubusercontent.com/4561443/197577861-19314e24-6b49-4bcf-9d7b-498b920e7d9d.png)


As you can see from these two graphs, the majority of the bytes being sent and received are on the `DataChannel`, the `MempoolChannel`, and the `VoteChannel`. 

The `VoteChannel` sends only `VoteMessages`:
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/consensus/reactor.go#L1076

The `MempoolChannel` sends `Txs`, lists of transactions, only:
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/mempool/v1/reactor.go#L238
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/mempool/v0/reactor.go#L239

The `DataChannel` communicates 3 message types,
`BlockPartMessages`: 
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/consensus/reactor.go#L529
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/consensus/reactor.go#L636
`ProposalMessages`:
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/consensus/reactor.go#L581
and `ProposalPOLMessages`:
https://github.com/tendermint/tendermint/blob/7a2e44184b219d99a6d3ae6c5f078ddc5f30f5e1/consensus/reactor.go#L593

**Comment by jmalicevic** [comment]
As for the mempool Txs, from the code the condition to not send a Tx to a peer is that we did not receive this transaction from a peer. We loop through the transactions in the pool and pick one by one, checking that it was not received from this peer. I don't know if the loop goes around the list once it reaches the end. If so, then we might be sending the same transaction multiple times.

**Comment by jmalicevic** [comment]
Btw, if we add the syncing protocols to the mix, we noticed that during statesync and blocksync, there are consensus messages being delivered as well. Maybe it does play a huge role in terms of % of bandwidth consumed but something to be considered.

**Comment by cason** [comment]
> It could be because (a) they are carrying (up to) 200 signatures each, 

A vote message carries a single signature. A block contains between 2/3+ and all signatures of Precommits of the previous height.

**Comment by cason** [comment]
> Similarly, are we frequently retransmitting the vote messages to peers? Are we tracking which peer has received a vote from us correctly and are we transmitting vote messages to peers that we already have evidence to believe do not need them?

There is a `HasVote` message, which is produced by the consensus protocol each time a vote is "accepted" (i.e., added to the list of valid received votes), and sent to every peer. So, we not only keep track of votes we received/sent from/to each peer, but we also inform everyone about votes we have received.

The same does not happen with block part messages (which are much bigger), I really don't know why.

**Comment by cason** [comment]
See https://github.com/tendermint/tendermint/issues/627.

**Comment by sergio-mena** [comment]
> A vote message carries a single signature. A block contains between 2/3+ and all signatures of Precommits of the previous height.

Good catch, Thanks!
However, I think the reasoning still applies: take a "precommit phase" in a particular height and round. You will need at least "2/3 worth-of-voting-power" (and usually all) vote messages (as opposed to one vote message containing them all, as I was wrongly implying above) to be broadcast all across the network, so it is still a real possibility for the bandwidth to be negatively affected by this (and, if so, we'd need to reopen the BLS discussion).

**Comment by cason** [comment]
I understand your point and I agree that there are too many votes being relayed. But I am afraid that no matter how we do that, we still need all validators to send at least one vote message, either with a single or multiple aggregated signatures, per round step. So question here is not whether all votes should be sent/relayed, but how many times are we sending/receiving the same vote.

**Comment by ebuchman** [comment]
> but how many times are we sending/receiving the same vote.

Is it worth trying to setup a metric to actually count this? Ie. how many times we receive a vote we already have. Same for block parts. The HasVote is supposed to minimize the amount of times we receive the same vote but there could still be some amplification. For block parts there is no HasBlockPart but it probably makes a lot of sense to add one! Probably shouldn't be that difficult and could result in significant savings!

Years ago I worked on a change to gossip block parts "rarest first" to help improve bandwidth usage but we never merged it (https://github.com/tendermint/spec/issues/309)

**Comment by thanethomson** [comment]
This issue will be superseded by #9880.
